### PR TITLE
#1059B #1042B File archive management

### PR DESCRIPTION
**Pull request by CarlosNZ**
Fix https://github.com/openmsupply/conforma-server/discussions/1059 (detailed spec/discussion: https://github.com/openmsupply/conforma-server/discussions/1059)

Requires back-end branch from https://github.com/openmsupply/conforma-server/pull/1060

Quite a big face-lift and consolidation for the Snapshots page. It gives more useful stats about each snapshot, include the file size and the archive status:
Screenshot 2023-07-26 at 12 05 21 PM

**Archives**:
Screenshot 2023-07-26 at 11 39 14 AM

## Summary of changes
- Separate lists for normal Snapshot and "Archive" snapshots
- Show snapshot stats, including file size and archive info
- No need to add time-stamp to snapshot names -- it's automatically appended to downloads
- Creating new snapshot, can specify archive range (or "none" or "full")
- Manually trigger a file archive
- Removed the ability to overwrite a snapshot -- you have to create a new one (can use the same name though, but will have a different timestamp)
- Lots of internal logic to handle the back-end changes (see inline comments)

### CODE REVIEW

File: docs/internal/Element-Type-Specs.md [hunk 1]
```diff
@@ -265,6 +265,7 @@ _Multi-choice question, with one allowed option, displayed as Drop-down list (Co
 - **persistUserInput** / **ignoreNullDefault**: See [above](#input-params)
 - **search**: `boolean` (default: `false`) -- if `true`, the list of options can be searched and filtered by user
 - **optionsDisplayProperty**: If `options` (above) consists of an array of objects, this parameter specifies the field of each object to be displayed in the options list. For example, if `options` was a list of organisation objects (i.e. `{orgId, name, licenceNumber}`), you'd probably specify `name` as the `optionsDisplayProperty`. Note that even though one field is displayed to the user in the Dropdown list, the _entire_ selected object is saved as the selection. And if `optionsDisplayProperty` refers to a field that doesn't exist on the supplied object, the plugin will fail and show in error in the application.
+- **optionsDisplayExpression**: Similar to `optionsDisplayProperty`, but you can build a display string out of multiple object properties, using the same substitution/templating mechanism as the [listBuilder](#list-builder-ingredients-list) `displayFormat` property.
```
Comment (CarlosNZ) [inline docs/internal/Element-Type-Specs.md:268]: Unrelated to this PR

File: src/containers/TemplateBuilder/shared/OperationContextHelpers.tsx [hunk 2]
```diff
@@ -269,7 +269,7 @@ export const importTemplate: ImportTemplateHelper =
       data.append('file', file)
 
       const result = await safeFetch(
-        getServerUrl('snapshot', { action: 'upload', name: snapshotName }),
```
Comment (CarlosNZ) [inline src/containers/TemplateBuilder/shared/OperationContextHelpers.tsx:272]: No need to include name as a parameter -- we'll use the filename of the upload as the name (since the file was named as such when downloading)

File: src/formElementPlugins/dropdownChoice/src/ApplicationView.tsx [hunk 3]
```diff
@@ -3,11 +3,25 @@ import { Dropdown, Label } from 'semantic-ui-react'
...
-type StringOptions = { options: string[]; optionsDisplayProperty: never }
+type ObjectOptions = { options: ObjectOption[] } & RequireAtLeastOne<{
+  optionsDisplayProperty: string
+  optionsDisplayExpression: string
+}>
+type StringOptions = {
+  options: string[]
+  optionsDisplayProperty: never
+  optionsDisplayExpression: never
+}
```
Comment (CarlosNZ) [inline src/formElementPlugins/dropdownChoice/src/ApplicationView.tsx:24]: Not part of this PR -- ignore

**Review by fergie-nz — APPROVED** [review]
Looks good.

Great to have those small fixes and UI changes.
